webapp/channels/src/components/admin_console/ip_filtering/index.tsx

```diff
@@ -0,0 +1,265 @@
+import type {AllowedIPRange, FetchIPResponse} from '@mattermost/types/config';
+
+import {Client4} from 'mattermost-redux/client';
+
+import AdminHeader from 'components/widgets/admin_console/admin_header';
+
+import IPFilteringAddOrEditModal from './add_edit_ip_filter_modal';
+import DeleteConfirmationModal from './delete_confirmation';
+import EditSection from './edit_section';
+import EnableSectionContent from './enable_section';
+import {isIPAddressInRanges} from './ip_filtering_utils';
+import SaveConfirmationModal from './save_confirmation_modal';
+
+import SaveChangesPanel from '../team_channel_settings/save_changes_panel';
+
+import './ip_filtering.scss';
+
+const IPFiltering = () => {
+    const {formatMessage} = useIntl();
+    const [showAddModal, setShowAddModal] = useState(false);
+    const [editFilter, setEditFilter] = useState<AllowedIPRange | null>(null);
+    const [ipFilters, setIpFilters] = useState<AllowedIPRange[] | null>(null);
+    const [originalIpFilters, setOriginalIpFilters] = useState<AllowedIPRange[] | null>(null);
+    const [saveNeeded, setSaveNeeded] = useState(false);
+    const [filterToDelete, setFilterToDelete] = useState<AllowedIPRange | null>(null);
+    const [currentUsersIP, setCurrentUsersIP] = useState<string | null>(null);
+    const [currentIPIsInRange, setCurrentIPIsInRange] = useState<boolean>(false);
+    const [saveConfirmationModal, setSaveConfirmationModal] = useState<JSX.Element | null>(null);
+    const [saving, setSaving] = useState<boolean>(false);
+
+    const [filterToggle, setFilterToggle] = useState<boolean>(false);
+
+    useEffect(() => {
+        Client4.getIPFilters().then((res) => {
+            setIpFilters(res as AllowedIPRange[]);
+            setOriginalIpFilters(res as AllowedIPRange[]);
+        });
+
+        Client4.getCurrentIP().then((res) => {
+            setCurrentUsersIP((res as FetchIPResponse)?.IP);
+        });
+    }, []);
+
+    useEffect(() => {
+        if (ipFilters === null || originalIpFilters === null) {
+            return;
+        }
+
+        // Check if the ipFilters list differs from the originalIpFilters list
+        const haveFiltersChanged = JSON.stringify(ipFilters) !== JSON.stringify(originalIpFilters);
+        setSaveNeeded(haveFiltersChanged);
+    }, [ipFilters, originalIpFilters]);
+
+    useEffect(() => {
+        if (!filterToggle) {
+            setCurrentIPIsInRange(true);
+            return;
+        }
+        if (!ipFilters?.length) {
+            setCurrentIPIsInRange(true);
+            return;
+        }
+        setCurrentIPIsInRange(ipFilters !== null && currentUsersIP !== null && isIPAddressInRanges(currentUsersIP, ipFilters));
+    }, [ipFilters, currentUsersIP, filterToggle]);
+
+    useEffect(() => {
+        if (!ipFilters?.length) {
+            return;
+        }
+        setFilterToggle(ipFilters?.some((filter: AllowedIPRange) => filter.Enabled === true) ?? false);
```

Comment:
Same with `filterToggle` and `ipFilters`. Since they sort of depend on each other, it'll also avoid any potential desyncs or loops if we update `ipFilters` directly (instead of using `useEffect`s) when a filter is enabled/disabled or you click on the toggle button. That'll also prevent extra renders caused by one state change causing another state change

Reply:
I need to know the previous value of the filterToggle between renders so that if the number of filters drops from 1 to 0 I can still have the toggle enabled and show the zero state. That's basically what the if statement in this is doing - if there are no filters, allow the toggle to independently set its value. If there are filters, then the value is based on if they are disabled or not.

Part of this is a quirk of the UX vs backend. There's no global enable/disable in the backend, it's per row.

Reply:
That seems a bit weird to me, but I guess that sort of thing happens in frontend coding where the UX doesn't quite line up. That's fine by me then